cli/file_watcher.rs

```diff
@@ -11,14 +12,75 @@ use notify::Error as NotifyError;
 use notify::RecommendedWatcher;
 use notify::RecursiveMode;
 use notify::Watcher;
+use std::mem;
 use std::path::PathBuf;
 use std::pin::Pin;
+use std::sync::{mpsc, mpsc::Receiver};
+use std::time::{Duration, Instant};
 use tokio::select;
-use tokio::sync::mpsc;
+
+const DEBOUNCE_TIMEOUT_MS: Duration = Duration::from_millis(200);
 
 // TODO(bartlomieju): rename
 type WatchFuture = Pin<Box<dyn Future<Output = Result<(), AnyError>>>>;
 
+struct Debounce {
+  rx: Receiver<Result<NotifyEvent, AnyError>>,
+  debounce_time: Duration,
+  start_time: Option<Instant>,
+  last_event: Option<NotifyEvent>,
+}
+
+impl Debounce {
+  fn new(
+    rx: Receiver<Result<NotifyEvent, AnyError>>,
+    debounce_time: Duration,
+  ) -> Self {
+    Self {
+      rx,
+      debounce_time,
+      start_time: None,
+      last_event: None,
+    }
+  }
+}
+
+impl Stream for Debounce {
+  type Item = NotifyEvent;
+
+  /// Note that this never returns `Poll::Ready(None)`, which means that file watcher will be alive
+  /// until the Deno process is terminated.
+  fn poll_next(
+    self: Pin<&mut Self>,
+    cx: &mut Context,
+  ) -> Poll<Option<Self::Item>> {
+    let inner = self.get_mut();
+    if let Ok(Ok(event)) = inner.rx.try_recv() {
+      if matches!(event.kind, EventKind::Create(_) | EventKind::Modify(_) | EventKind::Remove(_))
+      {
+        inner.start_time = Some(Instant::now());
+        inner.last_event = Some(event);
+      }
+    }
+
+    match &inner.last_event {
+      Some(_)
+        if inner.start_time.map_or(false, |start_time| {
+          start_time.elapsed() >= inner.debounce_time
+        }) =>
+      {
+        inner.start_time = None;
+        let event = mem::take(&mut inner.last_event);
+        Poll::Ready(event)
+      }
+      _ => {
+        cx.waker().wake_by_ref();
+        Poll::Pending
+      }
```

Comment:
This may lead to a busy loop, so ideally we insert sleep or something here. But at the same time we want to receive events as soon as they are sent.
Any thoughts on this? @bartlomieju